vault/seal/ocikms/ocikms.go

```diff
@@ -0,0 +1,305 @@
+		k.authTypeAPIKey, err = strconv.ParseBool(authTypeAPIKeyStr)
+		if err != nil {
+			metrics.SetGauge(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("failed parsing authTypeAPIKey parameter: {{err}}", err)
+		}
+	}
+	if k.authTypeAPIKey {
+		k.logger.Info(fmt.Sprintf("using OCI KMS with user principal"))
+	} else {
+		k.logger.Info(fmt.Sprintf("using OCI KMS with instance principal"))
+	}
+
+	// Check and set OCI KMS crypto client
+	if k.cryptoClient == nil {
+		kmsCryptoClient, err := k.getOCIKMSClient()
+		if err != nil {
+			metrics.SetGauge(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("error initializing OCI KMS clients: {{err}}", err)
+		}
+
+		// KeyID validation by trying to generate a DEK
+		keyLength := 32
+		includePlaintextKey := true
+		generateKeyDetails := keymanagement.GenerateKeyDetails{
+			IncludePlaintextKey: &includePlaintextKey,
+			KeyId:               &k.keyID,
+			KeyShape: &keymanagement.KeyShape{
+				Algorithm: "AES",
+				Length:    &keyLength,
+			},
+		}
+		requestMetadata := k.getRequestMetadata()
+		dekInput := keymanagement.GenerateDataEncryptionKeyRequest{
+			GenerateKeyDetails: generateKeyDetails,
+			RequestMetadata:    requestMetadata,
+		}
+		generateDEKResponse, err := kmsCryptoClient.GenerateDataEncryptionKey(context.Background(), dekInput)
+		if err != nil || generateDEKResponse.Ciphertext == nil {
+			metrics.SetGauge(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("failed keyID validation: {{err}}", err)
+		}
+		k.logger.Info(fmt.Sprintf("successfully validated keyID"))
+
+		// Store client
+		k.cryptoClient = kmsCryptoClient
+	}
+
+	// Map that holds non-sensitive configuration info
+	sealInfo := make(map[string]string)
+	sealInfo[KMSConfigKeyID] = k.keyID
+	sealInfo[KMSConfigCryptoEndpoint] = k.cryptoEndpoint
+
+	return sealInfo, nil
+}
+
+// Build OCI KMS crypto client
+func (k *OCIKMSSeal) getOCIKMSClient() (*keymanagement.KmsCryptoClient, error) {
+	var cp common.ConfigurationProvider
+	var err error
+	if k.authTypeAPIKey {
+		cp = common.DefaultConfigProvider()
+	} else {
+		cp, err = auth.InstancePrincipalConfigurationProvider()
+		if err != nil {
+			return nil, errwrap.Wrapf("failed creating InstancePrincipalConfigurationProvider: {{err}}", err)
+		}
+	}
+
+	// Build crypto client
+	kmsCryptoClient, err := keymanagement.NewKmsCryptoClientWithConfigurationProvider(cp, k.cryptoEndpoint)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed creating NewKmsCryptoClientWithConfigurationProvider: {{err}}", err)
+	}
+
+	return &kmsCryptoClient, nil
+}
+
+// Request metadata includes retry policy
+func (k *OCIKMSSeal) getRequestMetadata() common.RequestMetadata {
+	if !KMSShouldRetryOperation {
+		noRetryPolicy := common.NoRetryPolicy()
+		return common.RequestMetadata{
+			RetryPolicy: &noRetryPolicy,
+		}
+	}
+
+	// Only retry for 5xx errors
+	retryOn5xxFunc := func(r common.OCIOperationResponse) bool {
+		return !(r.Error == nil && r.Response.HTTPResponse().StatusCode < 500)
```

Comment:
```suggestion
		return r.Error != nil && r.Response.HTTPResponse().StatusCode >= 500
```
I think the non-negated version is more intuitive to read.

Reply:
Should it be || instead of && in your suggestion?